### PR TITLE
Add code for CMS signed object verification to cert_checker

### PR DESCRIPTION
**Pull request by Martin2112**
Add code to support draft RFC section 3.1 requires CMS signing of pre-certificates. This provides for checking that such objects are signed by a particular cert (which must be verified separately) and unpacking a DER encoded certificate payload from the signed object.

Also includes tests + supporting data plus some style guide related fixes.

### CODE REVIEW

**Comment by Martin2112** [comment]
Begging for a review of this PR again!

File: cpp/log/cert_checker.cc [hunk 1]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
...
+
+  sk_X509_free(validation_chain);
+
+  CMS_ContentInfo_free(cms_content_info);
+
+  return (verified == 1) ? util::Status::OK
+                         : util::Status(util::error::INVALID_ARGUMENT,
+                                        "CMS verification failed");
+}
+
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: extra blank line req'd

File: cpp/log/cert_checker.cc [hunk 2]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
...
+Cert* CertChecker::UnpackCmsSignedCertificate(BIO* cms_bio_in,
+                                              const Cert& verify_cert) {
+  BIO* unpacked_bio = BIO_new(BIO_s_mem());
+  Cert* const cert = new Cert();
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]:
```C++
unique_ptr<Cert> cert(new Cert);
...
return cert.release();
```

File: cpp/log/cert_checker.cc [hunk 3]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
...
+
+  CMS_ContentInfo_free(cms_content_info);
+
+  return (verified == 1) ? util::Status::OK
+                         : util::Status(util::error::INVALID_ARGUMENT,
+                                        "CMS verification failed");
+}
+
+Cert* CertChecker::UnpackCmsSignedCertificate(BIO* cms_bio_in,
+                                              const Cert& verify_cert) {
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: CHECK_NOTNULL(cms_bio_in);

File: cpp/log/cert_checker.cc [hunk 4]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
...
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+
+util::Status CertChecker::UnpackCmsDerBio(BIO* cms_bio_in,
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: `CHECK_NOTNULL` on the ptrs?

File: cpp/log/cert_checker.cc [hunk 5]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
   return false;
 }
 
+util::StatusOr<bool> CertChecker::IsCmsSignedByCert(CMS_ContentInfo* const cms,
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: `CHECK_NOTNULL(cms)` ?

File: cpp/log/cert_checker.cc [hunk 6]
```diff
@@ -359,4 +362,104 @@ StatusOr<bool> CertChecker::IsTrusted(const Cert& cert,
...
+    util::Status status = cert->LoadFromDerBio(unpacked_bio);
+
+    if (!status.ok()) {
+      LOG(WARNING) << "Could not unpack cert from CMS DER encoded data";
+    }
+  } else {
+    LOG_OPENSSL_ERRORS(ERROR);
+  }
+
+  BIO_free(cms_bio_in);
```
Comment (AlCutter) [inline cpp/log/cert_checker.cc]: This feels a bit naughty, shouldn't the caller do this?

File: cpp/log/cert_checker_test.cc [hunk 7]
```diff
@@ -131,6 +152,24 @@ class CertCheckerTest : public ::testing::Test {
   }
 };
 
+BIO* OpenTestFileBio(const string& filename) {
+  BIO* der_bio = BIO_new_file(filename.c_str(), "r");
+
+  CHECK(der_bio != nullptr);
```
Comment (AlCutter) [inline cpp/log/cert_checker_test.cc]: `CHECK_NOTNULL(der_bio);`

File: cpp/log/cert_checker_test.cc [hunk 8]
```diff
@@ -415,6 +454,123 @@ TEST_F(CertCheckerTest, ResolveIssuerCollisions) {
   EXPECT_OK(checker_.CheckCertChain(&chain2));
 }
 
+TEST_F(CertCheckerTest, CmsSignTestCase2) {
+  // In this test the embedded data is not a certificate in DER format
+  // but it doesn't get unpacked and the signature is valid.
+  Cert ca(ca_pem_);
+
+  CMS_ContentInfo* cms = ReadCmsFromDerFile(cert_dir_v2_ + kCmsSignedDataTest2);
+  ASSERT_TRUE(cms != nullptr);
```
Comment (AlCutter) [inline cpp/log/cert_checker_test.cc]: `ASSERT_NE(..)` ?  (elsewhere too)

File: cpp/log/cert_checker.cc (hunk 1 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 2 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 3 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 4 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done for input bio. Output can be nullptr if you want to verify but not unpack. It's documented in .h file.

File: cpp/log/cert_checker.cc (hunk 5 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 6 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: It's arguable but you might be right. I'll have another think about it.

File: cpp/log/cert_checker_test.cc (hunk 7 above)
Comment (Martin2112) [inline cpp/log/cert_checker_test.cc]: Done.

File: cpp/log/cert_checker_test.cc (hunk 8 above)
Comment (Martin2112) [inline cpp/log/cert_checker_test.cc]: Done.

File: cpp/log/cert_checker.cc (hunk 6 above)
Comment (Martin2112) [inline cpp/log/cert_checker.cc]: OK, we'll make the caller free it. I've updated the api and doc in header.

**Comment by Martin2112** [comment]
I think I've address all the comments + rebased to master.

**Comment by Martin2112** [comment]
Leak fixed. PTAL.

**Pull request merged**